cadc-dali/src/main/java/ca/nrc/cadc/dali/tables/parquet/ParquetWriter.java

```diff
@@ -0,0 +1,306 @@
+
+            if ("timestamp".equals(field.xtype)) {
+                fields.set(i, new VOTableField(field.getName(), "long"));
+            }
+
+            if ("short".equals(field.getDatatype())) {
+                fields.set(i, new VOTableField(field.getName(), "int", field.getArraysize()));
+            }
+        }
+
+        // Clear table data for empty VOTable
+        voTableResource.getTable().setTableData(null);
+    }
+
+    private static Map<String, String> prepareCustomMetaData(VOTableDocument voTableDocument, Long maxRec) throws IOException {
+        StringWriter stringWriter = new StringWriter();
+        VOTableWriter votableWriter = new VOTableWriter();
+        votableWriter.write(voTableDocument, stringWriter, maxRec);
+
+        Map<String, String> customMetaData = new HashMap<>();
+        customMetaData.put(IVOA_VOTABLE_PARQUET_VERSION_KEY, IVOA_VOTABLE_PARQUET_VERSION_VALUE);
+        customMetaData.put(IVOA_VOTABLE_PARQUET_CONTENT_KEY, stringWriter.toString());
+
+        return customMetaData;
+    }
+
+    private static int saveRecords(Long maxRec, TableData tableData, Schema schema, org.apache.parquet.hadoop.ParquetWriter<GenericRecord> writer)
+            throws IOException {
+        Iterator<List<Object>> iterator = tableData.iterator();
+        int recordCount = 1;
+
+        while (iterator.hasNext() && recordCount <= maxRec) {
+            GenericRecord record = new GenericData.Record(schema);
+            List<Object> rowData = iterator.next();
+
+            for (Schema.Field field : schema.getFields()) {
+                String columnName = field.name();
+                Schema unionSchema = field.schema().getTypes().get(1);
+                Object data = rowData.get(field.pos());
+
+                if (unionSchema.getType().equals(Schema.Type.ARRAY)) {
+                    String xtype = unionSchema.getProp("xtype");
+                    handleArrays(field, xtype, record, data);
+                } else if (unionSchema.getType().equals(Schema.Type.LONG)
+                        && (unionSchema.getLogicalType() != null
+                        && unionSchema.getLogicalType().getName().equals("timestamp-millis"))) {
+                    handleDateAndTimestamp(field, rowData, record, columnName);
+                } else {
+                    // handle primitives
+                    record.put(columnName, rowData.get(field.pos()));
+                }
+            }
+            writer.write(record);
+            recordCount++;
+        }
+        return recordCount;
+    }
+
+    private static void handleArrays(Schema.Field field, String xtype, GenericRecord record, Object data) {
+        if (xtype == null) {
+            record.put(field.name(), data);
+        } else {
+            handleXTypeArrayData(record, field.name(), xtype, data);
+        }
+    }
+
+    private static void handleXTypeArrayData(GenericRecord record, String columnName, String xtype, Object data) {
```

Comment:
Needing to use this much reflection is a bad sign and I think this could be improved, but not now: I have some ideas to improve/simplify Interval (in my CAOM 2.5 branch) and that might help simplify this. Nothing to do for now - this will work/fail correctly as-is.